lib/queuePopulator/IngestionProducer.js

```diff
@@ -185,95 +251,135 @@ class IngestionProducer {
     }
 
     /**
-     * get the list of objects for each bucket
+     * Get the list of object versions for a bucket
      *
-     * @param {object} bucketList - list of buckets
+     * @param {string} bucket - bucket name
+     * @param {object} state - previous state used to paginate version listing
+     * @param {string} [state.versionMarker] - NextVersionIdMarker
+     * @param {string} [state.keyMarker] - NextKeyMarker
      * @param {function} done - callback function
-     * @return {object} list of buckets and list of objects for each bucket
+     * @return {object} list of objects for each bucket, including a duplicate
+     *   entry for IsLatest versions
      */
-    _getBucketObjects(bucketList, done) {
-        if (!bucketList) {
-            return done(null, null);
+    _getObjectVersionsList(bucket, state, done) {
+        if (bucket === constants.usersBucket ||
+            bucket === constants.metastoreBucket) {
+            return done();
         }
-        return async.mapLimit(bucketList, 10, (bucketInfo, cb) => {
-            if (bucketInfo === constants.usersBucket ||
-            bucketInfo === constants.metastoreBucket) {
-                return cb();
+        const { versionMarker, keyMarker } = state;
+        const params = {
+            Bucket: bucket,
+        };
+        // if previous state, should paginate here
+        if (versionMarker && keyMarker) {
+            params.VersionIdMarker = versionMarker;
+            params.KeyMarker = keyMarker;
+        }
+        // TODO: For testing, I can set MaxKeys
+        const req = this._s3Client.listObjectVersions(params);
+        attachReqUids(req, this.requestLogger);
+        return req.send((err, data) => {
+            if (err) {
+                this.log.error('error getting list of object versions', {
+                    method: 'IngestionProducer._getObjectVersionsList',
+                    error: err,
+                    bucket,
+                });
+                return done(err);
             }
-            const req = this.ringReader.getObjectList({
-                Bucket: bucketInfo,
-            });
-            attachReqUids(req, this.requestLogger);
-            return req.send((err, data) => {
-                if (err) {
-                    this.log.error('error getting list of objects', {
-                        method: 'IngestionProducer:getBucketObjects', err });
-                    return cb(err);
-                }
-                return cb(null, { bucket: bucketInfo, objects: data.Contents });
-            });
-        }, (err, buckets) => done(err, buckets));
+            const {
+                IsTruncated,
+                NextKeyMarker,
+                NextVersionIdMarker,
+                Versions,
+                DeleteMarkers
+            } = data;
+
+            const versionList = [...Versions, ...DeleteMarkers];
+            // find all IsLatest versions since we need these as separate
```

Comment:
I should clean this up, I agree it's pretty weird logic.
Here, I wanted to accumulate all entries into a single list, including the master version, which I indicate by duplicating the `IsLatest` versions and putting a flag on them for the next method to process.

In addition, I realize it's probably best to not duplicate entry here, but after fetching object metadata (on next method) for the `IsLatest` version, I should create the duplicate entry there. It is possible that the `IsLatest` version and master version object metadata fetch might differ otherwise.

I'll look into removing the logic here, and creating the duplicate entry after fetching object metadata for `IsLatest` versions on the next method (`_getBucketObjectsMetadata`)